src/gas-price/gas-price.ts

```diff
@@ -0,0 +1,236 @@
+ * @returns
+ */
+export const calculateScalingMultiplier = (
+  recommendedGasPriceMultiplier: number,
+  scalingMultiplier: number,
+  lag: number,
+  scalingWindow: number
+) => recommendedGasPriceMultiplier + (scalingMultiplier - recommendedGasPriceMultiplier) * (lag / scalingWindow);
+
+/**
+ * Fetches the provider recommended gas price and saves it in the store.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @returns {ethers.BigNumber}
+ */
+export const updateGasPriceStore = async (chainId: string, providerName: string, rpcUrl: string) => {
+  const provider = new ethers.providers.StaticJsonRpcProvider(rpcUrl, {
+    chainId: Number.parseInt(chainId, 10),
+    name: chainId,
+  });
+
+  // Get the provider recommended gas price
+  const goGasPrice = await go(async () => provider.getGasPrice(), { retries: 1, attemptTimeoutMs: 2000 });
+  if (!goGasPrice.success) {
+    // eslint-disable-next-line no-console
+    console.log(`Failed to get provider gas price. Error: ${goGasPrice.error.message}.`);
+    throw goGasPrice.error;
+  }
+
+  // Save the new provider recommended gas price to the state
+  setStoreGasPrices(chainId, providerName, goGasPrice.data);
+
+  return goGasPrice.data;
+};
+
+/**
+ * Fetches the provider recommended gas price, saves it in the store and clears out expired gas prices.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ */
+export const gasPriceCollector = async (
+  chainId: string,
+  providerName: string,
+  rpcUrl: string,
+  sanitizationSamplingWindow: number
+) => {
+  // Initialize the gas store for the chain if not already present
+  initializeGasStore(chainId, providerName);
+  clearExpiredStoreGasPrices(chainId, providerName, sanitizationSamplingWindow);
+  await updateGasPriceStore(chainId, providerName, rpcUrl);
+};
+
+/**
+ *  Calculates the gas price to be used in a transaction based on sanitization and scaling settings.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @param gasSettings
+ * @param nonce
+ * @returns {ethers.BigNumber}
+ */
+export const airseekerV2ProviderRecommendedGasPrice = async (
+  chainId: string,
+  providerName: string,
+  rpcUrl: string,
+  gasSettings: GasSettings,
+  newDataFeedUpdateOnChainValues?: {
+    dataFeedId: string;
+    newDataFeedValue: DataFeedValue;
+  }
+): Promise<ethers.BigNumber> => {
+  const { recommendedGasPriceMultiplier, sanitizationPercentile, scalingWindow, scalingMultiplier } = gasSettings;
+  // Get the configured percentile of historical gas prices before adding the new price
+  const percentileGasPrice = getPercentile(
+    sanitizationPercentile,
+    gasPriceStore[chainId]![providerName]!.gasPrices.map((gasPrice) => gasPrice.price)
+  );
+
+  const gasPrice = await updateGasPriceStore(chainId, providerName, rpcUrl);
+  if (!gasPriceStore[chainId]![providerName]!.gasPrices) {
+    return multiplyGasPrice(gasPrice, recommendedGasPriceMultiplier);
+  }
+
+  const lastDataFeedValue =
```

Comment:
I thought we decided to use nonces to track updates.

Reply:
It's not guaranteed to work if we use an Airseeker with a single sponsorWallet for multiple datafeeds (and Burak requested that it would work for the v1 way in [this thread](https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1697721456428479?thread_ts=1697707242.099579&cid=C05TQPT7PNJ))